src/test/rpc/AccountSet_test.cpp

```diff
@@ -203,24 +206,164 @@ class AccountSet_test : public beast::unit_test::suite
+            jt[sfTransferRate.fieldName] = xfer_rate;
+            env(jt);
+            BEAST_EXPECT((*env.le(bob))[ sfTransferRate ] == xfer_rate);
+
+            // set a bad value over limit (> 2 * QUALITY_ONE)
+            xfer_rate = std::numeric_limits<std::uint32_t>::max();
+            jt[sfTransferRate.fieldName] = xfer_rate;
+            env(jt, ter(temBAD_TRANSFER_RATE));
+
+            // set a bad value over limit (> 2 * QUALITY_ONE)
+            xfer_rate = (2 * QUALITY_ONE) + 1;
+            jt[sfTransferRate.fieldName] = xfer_rate;
+            env(jt, ter(temBAD_TRANSFER_RATE));
+        }
+    }
+
+    void testGateway()
+    {
         using namespace test::jtx;
-        Env env(*this);
-        Account const alice ("alice");
-        env.fund(XRP(10000), alice);
-        auto jt = noop(alice);
+        auto runTest = [&](Env& env, double tr)
+        {
+            Account const alice ("alice");
+            Account const bob ("bob");
+            Account const gw ("gateway");
+            auto const USD = gw["USD"];
 
-        uint32 xfer_rate = 2000000000;
-        jt[sfTransferRate.fieldName] = xfer_rate;
-        env(jt);
-        BEAST_EXPECT((*env.le(alice))[ sfTransferRate ] == xfer_rate);
+            env.fund(XRP(10000), gw, alice, bob);
+            env.trust(USD(3), alice, bob);
+            env(rate(gw, tr));
+            env.close();
 
-        jt[sfTransferRate.fieldName] = 0u;
-        env(jt);
-        BEAST_EXPECT(! env.le(alice)->isFieldPresent(sfTransferRate));
+            auto const amount = USD(1);
+            Rate const rate (tr * QUALITY_ONE);
+            auto const amountWithRate = toAmount<STAmount> (multiply(amount.value(), rate));
+
+            env(pay(gw, alice, USD(3)));
+            env(pay(alice, bob, USD(1)), sendmax(USD(3)));
+
+            env.require(balance(alice, USD(3) - amountWithRate));
+            env.require(balance(bob, USD(1)));
+        };
 
-        // set a bad value (< QUALITY_ONE)
-        jt[sfTransferRate.fieldName] = 10u;
-        env(jt, ter(temBAD_TRANSFER_RATE));
+        // Test gateway with allowed transfer rates
+        {
+            Env env (*this);
+            runTest (env, 1.02);
+        }
+        {
+            Env env (*this);
+            runTest (env, 1);
+        }
+        {
+            Env env (*this);
+            runTest (env, 2);
+        }
+        {
+            Env env (*this);
+            runTest (env, 2.1);
+        }
+        {
+            Env env (*this, features(fix1201));
+            runTest (env, 1.02);
+        }
+        {
+            Env env (*this, features(fix1201));
+            runTest (env, 2);
+        }
+
+        // Test gateway when amendment is set after transfer rate
+        {
+            Env env (*this);
+            Account const alice ("alice");
+            Account const bob ("bob");
+            Account const gw ("gateway");
+            auto const USD = gw["USD"];
+            double tr = 2.75;
```

Comment:
I think leaving `USD` as `auto const` is fine. The exact details of the type in that instance aren't important.